src/main/java/streams/cta/stereo/Stereo.java

```diff
@@ -0,0 +1,296 @@
+
+    /**
+     * Estimate the position x,y of the showers impact from a collection of Plane objects
+     * @param planes planes for each telescope in the event
+     * @return an array [x, y] giving a point on the surface
+     */
+    double[] estimateCorePosition(List<Plane> planes){
+        int n = planes.size();
+
+        double[][] mat = new double[n][2];
+        double[] d =  new double[n];
+
+        for (int i = 0; i < n; i++) {
+            Plane plane = planes.get(i);
+            Vector3D norm = plane.getNormalAsVector();
+            double weight = plane.weight;
+
+            Vector2D projection = new Vector2D(norm.getX(), norm.getY());
+
+            mat[i][0] = projection.getX() * weight;
+            mat[i][1] = projection.getY() * weight;
+
+
+            Vector2D telPos = new Vector2D(plane.telescopePosition[0], plane.telescopePosition[1]);
+
+            d[i] = (projection.scalarMultiply(weight)).dotProduct(telPos);
+        }
+
+        //Do a linear least square regresssion
+        RealMatrix A = MatrixUtils.createRealMatrix(mat);
+        return MatrixUtils.inverse(A.transpose().multiply(A)).multiply(A.transpose()).operate(d);
+    }
+
+    /**
+     * Estimate the direction of the shower from a collection of Plane objects
+     * Returns a direction vector [x, y, z] pointing into the direction the shower
+     * originated.
+     * It does so by calculating a weighted sum of the estimated directions
+     * from all pairs of planes in the given as the method argument.
+     *
+     * @param planes the plane objects for each Telescope
+     * @return an array [x, y, z]
+     */
+    double[] estimateDirection(List<Plane> planes){
+
+        // get all combinations of size 2 in a rather inelegant way.
+        List<List<Plane>> tuples = new ArrayList<>();
+        for(Plane p1 : planes){
+            for(Plane p2 : planes){
+                if(p1 != p2){
+                    tuples.add(Lists.newArrayList(p1, p2));
+                }
+            }
+        }
+
+        Optional<Vector3D> direction = tuples.stream()
+                .map(l -> {
+
+                    Plane plane1 = l.get(0);
+                    Plane plane2 = l.get(1);
+                    double[] v1 = plane1.normalVector;
+                    double[] v2 = plane2.normalVector;
+
+                    Vector3D product = Vector3D.crossProduct(new Vector3D(v1), new Vector3D(v2));
+
+                    //dont know what happens now. here is the docstring from the python
+                    // # two great circles cross each other twice (one would be
+                    // # the origin, the other one the direction of the gamma) it
+                    // # doesn't matter which we pick but it should at least be
+                    // # consistent: make sure to always take the "upper"
+                    // # solution
+                    if (product.getZ() < 0) {
+                        product = product.scalarMultiply(-1);
+                    }
```

Comment:
does it make sense to mark this as #TODO in order to find the meaning of this code later?